controllers/storageautoscaler/storageautoscaler_controller.go

```diff
@@ -0,0 +1,202 @@
+	OperatorNamespace string
+	SyncMap           *sync.Map
+	EventCh           chan event.GenericEvent
+}
+
+// SetupWithManager sets up the reconciler with the manager
+func (r *StorageAutoscalerReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	// get the eventCh from the storage autoscaler scraper
+	storageAutoscalerReconcilerController := ctrl.NewControllerManagedBy(mgr).
+		For(&ocsv1.StorageAutoScaler{}, builder.WithPredicates(predicate.GenerationChangedPredicate{})).
+		// watch for generic events to trigger the reconcile
+		WatchesRawSource(source.Channel(r.EventCh,
+			&handler.EnqueueRequestForObject{},
+		))
+
+	return storageAutoscalerReconcilerController.Complete(r)
+}
+
+// +kubebuilder:rbac:groups="monitoring.coreos.com",resources=*,verbs=get
+
+// Reconcile reconciles the StorageAutoscaler object
+func (r *StorageAutoscalerReconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	r.Log.Info("reconciling storage autoscaler", "namespace", request.Namespace, "name", request.Name)
+
+	// list the storage autoscaler reconciler object
+	storageAutoScaler := &ocsv1.StorageAutoScaler{}
+	err := r.Get(ctx, client.ObjectKey{Namespace: request.Namespace, Name: request.Name}, storageAutoScaler)
+	if err != nil {
+		r.Log.Error(err, "failed to get storage autoscaler")
+		return reconcile.Result{}, err
+	}
+
+	// if no expansion is in-progress update the phase to "not-started"
+	if storageAutoScaler.Status.Phase == "" {
+		storageAutoScaler.Status.Phase = "NotStarted"
+		err := r.updateStatus(ctx, storageAutoScaler)
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+	}
+
+	// if the expansion is in-progress, verify the scaling
+	if storageAutoScaler.Status.Phase == "InProgress" {
+		return r.verifyScaling(ctx, storageAutoScaler)
+	}
+
+	// list the storagecluster
+	storageCluster := &ocsv1.StorageCluster{}
+	storageClusterName := storageAutoScaler.Spec.StorageCluster.Name
+	err = r.Get(ctx, client.ObjectKey{Name: storageClusterName, Namespace: request.Namespace}, storageCluster)
+	if err != nil {
+		r.Log.Error(err, "failed to get storage cluster")
+		return reconcile.Result{}, err
+	}
+
+	// list the storagecluster devicset storageclass
+	err = r.detectLsoStorageclass(ctx, storageCluster)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// get the osd usage from the sync map
+	usage, ok := r.SyncMap.Load(storageAutoScaler.Spec.DeviceClass)
+	if !ok {
+		err = fmt.Errorf("osd usage not found for device class")
+		r.Log.Error(err, "device class", storageAutoScaler.Spec.DeviceClass)
+		return reconcile.Result{}, err
+	}
+
+	if usage == nil {
+		err = fmt.Errorf("osd usage is nil for device class")
+		r.Log.Error(err, "device class", storageAutoScaler.Spec.DeviceClass)
+		return reconcile.Result{}, err
+	}
+
+	// check if the highest osd usage is greater than the threshold
+	if !((usage.(float64) * 100) >= float64(storageAutoScaler.Spec.StorageScalingThresholdPercent)) {
```

Comment:
A simple helper method would be good for this calculation so it can be unit tested